src/rdo.rs

```diff
@@ -77,6 +91,86 @@ pub struct RDOPartitionOutput {
+  rate_bins: Vec<Vec<Vec<u64>>>,
+  rate_counts: Vec<Vec<Vec<u64>>>,
+}
+
+impl RDOTracker {
+  pub fn new() -> RDOTracker {
+    RDOTracker {
+      rate_bins: vec![vec![vec![0; RDO_NUM_BINS]; TxSize::TX_SIZES_ALL]; RDO_QUANT_BINS],
+      rate_counts: vec![vec![vec![0; RDO_NUM_BINS]; TxSize::TX_SIZES_ALL]; RDO_QUANT_BINS],
+    }
+  }
+  fn merge_array(new: &mut Vec<u64>, old: &Vec<u64>) {
+    for (n, o) in new.iter_mut().zip(old.iter()) {
+      *n += o;
+    }
+  }
+  fn merge_2d_array(new: &mut Vec<Vec<u64>>, old: &Vec<Vec<u64>>) {
+    for (n, o) in new.iter_mut().zip(old.iter()) {
+      RDOTracker::merge_array(n, o);
+    }
+  }
+  fn merge_3d_array(new: &mut Vec<Vec<Vec<u64>>>, old: &Vec<Vec<Vec<u64>>>) {
+    for (n, o) in new.iter_mut().zip(old.iter()) {
+      RDOTracker::merge_2d_array(n, o);
+    }
+  }
+  pub fn merge_in(&mut self, input: &RDOTracker) {
+    RDOTracker::merge_3d_array(&mut self.rate_bins, &input.rate_bins);
+    RDOTracker::merge_3d_array(&mut self.rate_counts, &input.rate_counts);
+  }
+  pub fn add_rate(&mut self, qindex: u8, ts: TxSize, fast_distortion: u64, rate: u64) {
+    if fast_distortion != 0 {
+      let bs_index = ts as usize;
+      let q_bin_idx = (qindex as usize)/RDO_QUANT_DIV;
+      let bin_idx_tmp = ((fast_distortion as i64 - (RATE_EST_BIN_SIZE as i64) / 2) as u64 / RATE_EST_BIN_SIZE) as usize;
+      let bin_idx = if bin_idx_tmp >= RDO_NUM_BINS {
+        RDO_NUM_BINS - 1
+      } else {
+        bin_idx_tmp
+      };
+      self.rate_counts[q_bin_idx][bs_index][bin_idx] += 1;
+      self.rate_bins[q_bin_idx][bs_index][bin_idx] += rate;
+    }
+  }
+  pub fn print_code(&self) {
+    println!("pub static RDO_RATE_TABLE: [[[u64; RDO_NUM_BINS]; TxSize::TX_SIZES_ALL]; RDO_QUANT_BINS] = [");
+    for q_bin in 0..RDO_QUANT_BINS {
+      print!("[");
+      for bs_index in 0..TxSize::TX_SIZES_ALL {
+        print!("[");
+        for (rate_total, rate_count) in self.rate_bins[q_bin][bs_index].iter().zip(self.rate_counts[q_bin][bs_index].iter()) {
+          if *rate_count > 100 {
+            print!("{},", rate_total / rate_count);
+          } else {
+            print!("99999,");
+          }
+        }
+        println!("],");
+      }
+      println!("],");
+    }
+    println!("];");
+  }
+}
+
+pub fn estimate_rate(qindex: u8, ts: TxSize, fast_distortion: u64) -> u64 {
+  let bs_index = ts as usize;
+  let q_bin_idx = (qindex as usize)/RDO_QUANT_DIV;
+  let bin_idx_down = ((fast_distortion) / RATE_EST_BIN_SIZE).min((RDO_NUM_BINS - 2) as u64);
+  let bin_idx_up = (bin_idx_down + 1).min((RDO_NUM_BINS - 1) as u64);
+  let x0 = (bin_idx_down * RATE_EST_BIN_SIZE) as i64;
+  let x1 = (bin_idx_up * RATE_EST_BIN_SIZE) as i64;
+  let y0 = RDO_RATE_TABLE[q_bin_idx][bs_index][bin_idx_down as usize] as i64;
+  let y1 = RDO_RATE_TABLE[q_bin_idx][bs_index][bin_idx_up as usize] as i64;
+  let slope = ((y1 - y0) << 8) / (x1 - x0);
+  (y0 + (((fast_distortion as i64 - x0) * slope) >> 8)) as u64
```

Comment:
We admit extrapolating beyond the table range based on the final pair of each row. However, if these are decreasing then the estimate may be negative which is then cast to a very large value. Constraining the estimate to non-negative values resolves the overflows seen in tests.
```suggestion
  (y0 + (((fast_distortion as i64 - x0) * slope) >> 8)).max(0) as u64
```